src/main/scala/viper/server/vsi/Cache.scala

```diff
@@ -0,0 +1,204 @@
+  * its verifications (from a previous verification) must be stored in order to return them.
+  *
+  * As mentioned earlier, this requires an elaborate storage system that can
+  *
+  *   1) store verification results for individual program members
+  *   2) decide whether or not stored results are still valid if the program is changed.
+  *
+  * Once such a system is in place it can be used to implement a simple API. This API should take a
+  * program P, analyse it, transforms it and return a program P' as well as a set of verification results.
+  * This program should thereby have been transformed in such a way that a verifier may not perform
+  * unnecessary verifications on its members.
+  * */
+abstract class VerificationServerInterfaceCache {
+
+  override def toString: String = _cache.toString
+
+  /** The cache infrastructure is a nested, mutable Map. The maps can be described in terms of their
+    * key and values as follows:  (FileName -> (hashString -> cacheEntries)).
+    *
+    * The inner map is referred to as the fileCache. As the name indicates, it stores,
+    * for each file, a number of hashes and corresponding cache entries.
+    */
+  protected val _cache = MutableMap[String, FileCash]()
+  type FileCash = MutableMap[String, List[CacheEntry]]
+
+  /** This method transforms a program by applying cache's current state.
+    *
+    * The input and out put program should be equivalent with respect to verification, but they might
+    * differ in their AST. In particular the, output program should might be transformed such that
+    * verifying it is faster than verifying the input program.
+    *
+    * Additionally, the method returns previously cached verification result for members of the
+    * ast that hit the cache.
+    * */
+  def apply(
+        file_key: String,
+        input_prog: AST): (AST, List[CacheEntry]) = {
+
+    val cache_entries: ListBuffer[CacheEntry] = ListBuffer()
+    val concerningsToCache: ListBuffer[Concerning] = ListBuffer()
+    val concerningsToVerify: ListBuffer[Concerning] = ListBuffer()
+
+    //read errors from cache
+    val cachable_members = input_prog.decompose()
+    cachable_members.foreach((c: Concerning) => {
+      val dependencies = c.getDependencies(input_prog)
+      get(file_key, c, dependencies) match {
+        case Some(matched_entry) =>
+          concerningsToCache += c.transform
+          cache_entries += matched_entry
+        case None =>
+          //Nothing in cache, request verification
+          concerningsToVerify += c
+      }
+    })
+    val all_concernings: List[Concerning] = concerningsToCache.toList ++ concerningsToVerify.toList
+    val output_prog: AST = input_prog.compose(all_concernings)
+    (output_prog, cache_entries.toList)
+  }
+
+  /** Utility function to retrieve entries for single members.
+    * */
+  final def get(
+        file_key: String,
+        key: Concerning,
+        dependencies: List[Member]): Option[CacheEntry] = {
```

Comment:
Why is it necessary to ask for a ```dependencies``` parameter here? Could we not (lazily) compute it for the current AST?

Reply:
The same comment applies to the ```update``` method below.